simplestore/src/main/java/com/uber/simplestore/impl/AtomicFile.java

```diff
@@ -89,20 +92,55 @@ public FileOutputStream startWrite() throws IOException {
     if (mLegacyBackupName.exists()) {
       rename(mLegacyBackupName, mBaseName);
     }
+    return createFileOutputStreamForWrite(true);
+  }
+
+  private FileOutputStream createParentLegacy(FileNotFoundException e) throws IOException {
+    File parent = mNewName.getParentFile();
+    if (!parent.mkdirs()) {
+      throw new IOException("Failed to create directory for " + mNewName);
+    }
+    try {
+      return new FileOutputStream(mNewName);
+    } catch (FileNotFoundException e2) {
+      throw new IOException("Failed to create new file " + mNewName, e2);
+    }
+  }
 
+  @VisibleForTesting
+  FileOutputStream createFileOutputStreamForWrite(boolean fixOnFileNotFoundationException)
+      throws IOException {
     try {
       return new FileOutputStream(mNewName);
     } catch (FileNotFoundException e) {
-      File parent = mNewName.getParentFile();
-      if (!parent.mkdirs()) {
-        throw new IOException("Failed to create directory for " + mNewName);
-      }
-      try {
-        return new FileOutputStream(mNewName);
-      } catch (FileNotFoundException e2) {
-        throw new IOException("Failed to create new file " + mNewName, e2);
+      return createParent(e, fixOnFileNotFoundationException);
+    }
+  }
+
+  private FileOutputStream createParent(
+      FileNotFoundException rawError, boolean fixOnFileNotFoundationException) throws IOException {
+    if (fixOnFileNotFoundationException) {
+      return createParentWithFix(rawError);
+    } else {
+      return createParentLegacy(rawError);
+    }
+  }
+
+  private FileOutputStream createParentWithFix(FileNotFoundException rawError) throws IOException {
+    File parent = mNewName.getParentFile();
+    if (parent == null) {
+      throw rawError;
+    }
+    synchronized (lock) {
+      if (!parent.exists() && !parent.mkdirs()) {
+        throw rawError;
       }
```

Comment:
Those are the core code to make the fix effective. One should be noted that `synchronized` will only be applied when the folder was missing. If the folder is already present, then those piece of code will no longer be executed again.